doc/libserial.2
.\" man page for Munts Technologies Linux Simple I/O Library
.\"
.\" Copyright (C)2016-2025, Philip Munts dba Munts Technologies.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\" * Redistributions of source code must retain the above copyright notice,
.\"   this list of conditions and the following disclaimer.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.TH libserial 2 "25 September 2025" "version 1" "Linux Simple I/O Library"
.SH NAME
.B libserial
\-\- Linux Simple I/O Library: Asynchronous Serial Port Module
.SH SYNOPSIS
.nf
.B #include <libsimpleio/libserial.h>

.B typedef enum
.B {
.B "  SERIAL_PARITY_NONE,"
.B "  SERIAL_PARITY_EVEN,"
.B "  SERIAL_PARITY_ODD,"
.B } SERIAL_PARITY_t;

.B typedef enum
.B {
.B "  SERIAL_FLUSH_INPUT,"
.B "  SERIAL_FLUSH_OUTPUT,"
.B "  SERIAL_FLUSH_BOTH ,"
.B } SERIAL_FLUSH_t;

.BI "void SERIAL_open(const char *" name ", int32_t " baudrate ", int32_t " parity ","
.BI "  int32_t " databits ", int32_t " stopbits ", int32_t *"fd ", int32_t *" error ");"

.BI "void SERIAL_close(int32_t " fd ", int32_t *" error ");"

.BI "void SERIAL_send(int32_t " fd ", void *" buf ", int32_t " bufsize ","
.BI "  int32_t *" count ", int32_t *" error ");"

.BI "void SERIAL_receive(int32_t " fd ", void *" buf ", int32_t " bufsize ","
.BI "  int32_t *" count ", int32_t *" error ");"

.BI "void SERIAL_flush(int32_t " fd ", int32_t " what ", int32_t *" error ");"

.fi
Link with
.BR -lsimpleio .
.SH DESCRIPTION
.nh
All functions return either
.B 0
(upon success) or an
.B errno
value (upon failure) in
.IR *error .
.PP
.B SERIAL_open()
opens and configures a serial port device.  The device name must be passed in the
.I name
parameter.
The
.I baudrate
parameter sets the serial port bit rate.  Allowed values are
.BR 50 ", " 75 ", " 110 ", " 134 ", " 150 ", " 200 ", " 300 ", " 600 ", "
.BR 1200 ", " 1800 ", " 2400 ", " 4800 ", " 9600 ", " 19200 ", " 38400 ", "
.BR 57600 ", " 115200 ", " 230400 ", " 460800 ", " 500000 ", " 576000 ", "
.BR 921600 ", and " 1000000 "."
The
.I parity
parameter sets the parity mode.  Allowed values are
.BR SERIAL_PARITY_NONE ", " SERIAL_PARITY_EVEN ", and " SERIAL_PARITY_ODD .
The
.I databits
parameter sets the number of bits per character, and may be from
.BR 5  " to " 8 .
The
.I stopbits
parameter sets the number of stop bits per character, and may be
.BR 1 " or " 2 .
Upon success, a file descriptor for the
serial port device is returned in
.IR *fd .
.PP
.B SERIAL_close()
closes a previously opened serial port device.
The serial port device file descriptor is passed in
.IR fd .
.P
.B SERIAL_send()
sends data to the serial port device.
The serial port device file descriptor is passed in
.IR fd .
The transmit buffer address is passed in
.IR buf
and its size is passed in
.IR bufsize .
Upon success, the number of bytes actually sent will be returned in
.IR *count .
.PP
.B SERIAL_receive()
receives data from the serial port device.
The serial port device file descriptor is passed in
.IR fd .
The receive buffer address is passed in
.IR buf
and its size is passed in
.IR bufsize .
Upon success, the number of bytes actually received will be returned in
.IR *count .
.PP
.B SERIAL_flush()
flushes the serial port queue(s) selected by
.IR what .
.SH SEE ALSO
.BR libsimpleio "(2), " libadc "(2), " libdac "(2), " libevent "(2), " libgpio "(2),"
.br
.BR libhidraw "(2), " libiio "(2), " libipv4 "(2), " libi2c "(2), " liblinux "(2),"
.br
.BR liblinx "(2), " libpwm "(2), " libspi "(2), " libstream "(2), " libwatchdog "(2)"
.SH AUTHOR
Philip Munts dba Munts Technologies.


c/libserial.c
/* Serial port device wrapper services for Linux */

// Copyright (C)2016-2025, Philip Munts dba Munts Technologies.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <termios.h>
#include <unistd.h>

#include "macros.inc"
#include "libserial.h"

void SERIAL_open(const char *name, int32_t baudrate, int32_t parity,
  int32_t databits, int32_t stopbits, int32_t *fd, int32_t *error)
{
  assert(error != NULL);

  struct termios cfg;

  // Validate parameters

  if (fd == NULL)
  {
    *error = EINVAL;
    ERRORMSG("fd argument is NULL", *error, __LINE__ - 3);
    return;
  }

  if (name == NULL)
  {
    *fd = -1;
    *error = EINVAL;
    ERRORMSG("name argument is NULL", *error, __LINE__ - 4);
    return;
  }

  switch (baudrate)
  {
    case 50 :
    case 75 :
    case 110 :
    case 134 :
    case 150 :
    case 200 :
    case 300 :
    case 600 :
    case 1200 :
    case 1800 :
    case 2400 :
    case 4800 :
    case 9600 :
    case 19200 :
    case 38400 :
    case 57600 :
    case 115200 :
    case 230400 :
#ifdef B460800
    case 460800 :
#endif
#ifdef B500000
    case 500000 :
#endif
#ifdef B576000
    case 576000 :
#endif
#ifdef B921600
    case 921600 :
#endif
#ifdef B1000000
    case 1000000 :
#endif
#ifdef B1152000
    case 1152000 :
#endif
#ifdef B1500000
    case 1500000 :
#endif
#ifdef B2000000
    case 2000000 :
#endif
#ifdef B2500000
    case 2500000 :
#endif
#ifdef B3000000
    case 3000000 :
#endif
#ifdef B3500000
    case 3500000 :
#endif
#ifdef B4000000
    case 4000000 :
#endif
      break;

    default :
      *fd = -1;
      *error = EINVAL;
      ERRORMSG("baudrate argument is invalid", *error, __LINE__ - 25);
      return;
  }

  if ((parity < SERIAL_PARITY_NONE) || (parity > SERIAL_PARITY_ODD))
  {
    *fd = -1;
    *error = EINVAL;
    ERRORMSG("parity argument is invalid", *error, __LINE__ - 4);
    return;
  }

  if ((databits < 5) || (databits > 8))
  {
    *fd = -1;
    *error = EINVAL;
    ERRORMSG("databits argument is invalid", *error, __LINE__ - 4);
    return;
  }

  if ((stopbits < 1) || (stopbits > 2))
  {
    *fd = -1;
    *error = EINVAL;
    ERRORMSG("stopbits argument is invalid", *error, __LINE__ - 4);
    return;
  }

  // Open serial port device

  *fd = open(name, O_RDWR|O_NOCTTY);
  if (*fd < 0)
  {
    *fd = -1;
    *error = errno;
    ERRORMSG("open() failed", *error, __LINE__ - 5);
    return;
  }

  // Configure serial port device

  if (tcgetattr(*fd, &cfg))
  {
    *error = errno;
    ERRORMSG("tcgetattr() failed", *error, __LINE__ - 3);
    close(*fd);
    *fd = -1;
    return;
  }

  // Put serial port into raw mode

  cfmakeraw(&cfg);

  cfg.c_cflag &= ~HUPCL;
  cfg.c_iflag = 0;
  cfg.c_oflag = 0;
  cfg.c_lflag = 0;

  // Set baud rate

  cfg.c_cflag &= ~CBAUD;

  switch (baudrate)
  {
    case 50:
      cfg.c_cflag |= B50;
      break;

    case 75:
      cfg.c_cflag |= B75;
      break;

    case 110:
      cfg.c_cflag |= B110;
      break;

    case 134:
      cfg.c_cflag |= B134;
      break;

    case 150:
      cfg.c_cflag |= B150;
      break;

    case 200:
      cfg.c_cflag |= B200;
      break;

    case 300:
      cfg.c_cflag |= B300;
      break;

    case 600:
      cfg.c_cflag |= B600;
      break;

    case 1200:
      cfg.c_cflag |= B1200;
      break;

    case 1800:
      cfg.c_cflag |= B1800;
      break;

    case 2400:
      cfg.c_cflag |= B2400;
      break;

    case 4800:
      cfg.c_cflag |= B4800;
      break;

    case 9600:
      cfg.c_cflag |= B9600;
      break;

    case 19200:
      cfg.c_cflag |= B19200;
      break;

    case 38400:
      cfg.c_cflag |= B38400;
      break;

    case 57600:
      cfg.c_cflag |= B57600;
      break;

    case 115200:
      cfg.c_cflag |= B115200;
      break;

    case 230400:
      cfg.c_cflag |= B230400;
      break;

#ifdef B460800
    case 460800:
      cfg.c_cflag |= B460800;
      break;
#endif

#ifdef B500000
    case 500000:
      cfg.c_cflag |= B500000;
      break;
#endif

#ifdef B576000
    case 576000:
      cfg.c_cflag |= B576000;
      break;
#endif

#ifdef B921600
    case 921600:
      cfg.c_cflag |= B921600;
      break;
#endif

#ifdef B1000000
    case 1000000:
      cfg.c_cflag |= B1000000;
      break;
#endif

#ifdef B1152000
    case 1152000:
      cfg.c_cflag |= B1152000;
      break;
#endif

#ifdef B1500000
    case 1500000:
      cfg.c_cflag |= B1500000;
      break;
#endif

#ifdef B2000000
    case 2000000:
      cfg.c_cflag |= B2000000;
      break;
#endif

#ifdef B2500000
    case 2500000:
      cfg.c_cflag |= B2500000;
      break;
#endif

#ifdef B3000000
    case 3000000:
      cfg.c_cflag |= B3000000;
      break;
#endif

#ifdef B350000
    case 3500000:
      cfg.c_cflag |= B3500000;
      break;
#endif

#ifdef B4000000
    case 4000000:
      cfg.c_cflag |= B4000000;
      break;
#endif
  }

  // Set parity mode

  switch (parity)
  {
    case SERIAL_PARITY_NONE :
      cfg.c_cflag &= ~PARENB;
      cfg.c_cflag &= ~PARODD;
      cfg.c_iflag &= ~INPCK;
      break;

    case SERIAL_PARITY_EVEN :
      cfg.c_cflag |= PARENB;
      cfg.c_cflag &= ~PARODD;
      cfg.c_iflag |= INPCK;
      break;

    case SERIAL_PARITY_ODD :
      cfg.c_cflag |= PARENB;
      cfg.c_cflag |= PARODD;
      cfg.c_iflag |= INPCK;
      break;
  }

  // Set data bits

  cfg.c_cflag &= ~CSIZE;

  switch (databits)
  {
    case 5 :
      cfg.c_cflag |= CS5;
      break;

    case 6 :
      cfg.c_cflag |= CS6;
      break;

    case 7 :
      cfg.c_cflag |= CS7;
      break;

    case 8 :
      cfg.c_cflag |= CS8;
      break;
  }

  // Set stop bits

  switch (stopbits)
  {
    case 2 :
      cfg.c_cflag |= CSTOPB;
      break;
  }

  if (tcsetattr(*fd, TCSANOW, &cfg))
  {
    *error = errno;
    ERRORMSG("tcgetattr() failed", *error, __LINE__ - 3);
    close(*fd);
    *fd = -1;
    return;
  }

  // Flush serial buffers

  usleep(100000); // Don't know why this delay is necessary

  if (tcflush(*fd, TCIOFLUSH) < 0)
  {
    *error = errno;
    ERRORMSG("tcflush() failed", *error, __LINE__ - 3);
    close(*fd);
    *fd = -1;
    return;
  }

  *error = 0;
}

static const int32_t QueueSelector[3] = { TCIFLUSH, TCOFLUSH, TCIOFLUSH };

void SERIAL_flush(int32_t fd, int32_t what, int32_t *error)
{
  assert(error != NULL);

  // Validate parameters

  if (fd < 0)
  {
    *error = EINVAL;
    ERRORMSG("fd argument is invalid", *error, __LINE__ - 3);
    return;
  }

  if ((what < SERIAL_FLUSH_INPUT) || (what > SERIAL_FLUSH_BOTH))
  {
    *error = EINVAL;
    ERRORMSG("what argument is invalid", *error, __LINE__ - 3);
    return;
  }

  if (tcflush(fd, QueueSelector[what]) < 0)
  {
    *error = errno;
    ERRORMSG("tcflush() failed", *error, __LINE__ - 3);
    return;
  }

  *error = 0;
}
